client/tests/util/getPage.js

```diff
@@ -0,0 +1,134 @@
+const baseUrl = `http://localhost:${CLIENT_PORT}`;
+
+const startServer = async serverOrClient => {
+    return new Promise(resolve => {
+        const server = spawn('yarn', ['workspace', serverOrClient, 'dev'], {
+            cwd: path.resolve(__dirname, '../../'),
+            env: {
+                PATH: process.env.PATH,
+                PORT,
+                CLIENT_PORT,
+                AUTOMATION_SCHEDULER_PORT,
+                API_SERVER: `http://localhost:${PORT}`,
+                APP_SERVER: baseUrl,
+                AUTOMATION_SCHEDULER_URL: `http://localhost:${AUTOMATION_SCHEDULER_PORT}`,
+                PGDATABASE: 'aria_at_report_test',
+                PGPORT: 5432,
+                ENVIRONMENT: 'test'
+            }
+        });
+
+        const killServer = async () => {
+            await new Promise((resolve, reject) => {
+                treeKill(server.pid, error => {
+                    if (error) return reject(error);
+                    resolve();
+                });
+            });
+        };
+
+        server.stdout.on('data', data => {
+            const output = data.toString();
+            console.info(output); // eslint-disable-line no-console
+
+            if (
+                (serverOrClient === 'server' &&
+                    output.includes(`Listening on ${PORT}`)) ||
+                (serverOrClient === 'client' &&
+                    output.includes('compiled successfully'))
+            ) {
+                resolve({ close: killServer });
+            }
+        });
+
+        server.stderr.on('data', data => {
+            const output = data.toString();
+            console.info(output); // eslint-disable-line no-console
+        });
+    });
+};
+
+const setup = async () => {
+    // eslint-disable-next-line no-console
+    console.info(
+        'Starting dev servers. This is required for end-to-end testing'
+    );
+    [clientServer, backendServer] = await Promise.all([
+        startServer('client'),
+        startServer('server')
+    ]);
+
+    browser = await puppeteer.launch({
+        headless: 'new',
+        args: ['--no-sandbox'] // Required for GitHub environment
+    });
+};
+
+const teardown = async () => {
+    await Promise.all([backendServer.close(), clientServer.close()]);
+
+    // Browser might not be defined, if it failed to start
+    if (browser) await browser.close();
+};
+
+let incognitoContexts = {};
+
+const getPage = async (options, callback) => {
+    const { role, url } = options;
+    if (role == null || !['admin', 'tester', 'vendor', false].includes(role)) {
+        throw new Error('Please provide a valid role');
+    }
+
+    if (!incognitoContexts[role]) {
+        incognitoContexts[role] = await browser.createIncognitoBrowserContext();
+    }
+    const incognitoContext = incognitoContexts[role];
+
+    const page = await incognitoContext.newPage();
+
+    if (!url) {
+        throw new Error('Please provide a URL, even if it it is simply "/"');
```

Comment:
Defaulting `url` to `/` and throwing a console.warn about `/` being used when defaulted seems fine to me. Unless there is another reason you may not want to do that?

Reply:
I like an error in this case because it prevents the url from being undefined. With a warning there wouldn't be any guarantee about whether url is undefined and I would have to keep track of that gotcha throughout the function.

Reply:
Okay got it